File: apps/agent/tools/ide_review/mcp_client.py
"""最小 MCP stdio JSON-RPC 客户端（无第三方依赖，仅供 M0 可选 live）。

支持 MCP 标准 Content-Length 帧；兼容部分实现的 NDJSON 行协议。
"""
from __future__ import annotations

import json
import os
import shutil
import subprocess
import threading
from typing import Any


class McpStdioError(RuntimeError):
    pass


class McpStdioClient:
    """极简 MCP client：initialize → tools/list → tools/call。"""

    def __init__(
        self,
        command: str,
        args: list[str],
        *,
        cwd: str | None = None,
        timeout_sec: float = 45.0,
    ) -> None:
        self.command = command
        self.args = args
        self.cwd = cwd
        self.timeout_sec = timeout_sec
        self._proc: subprocess.Popen[bytes] | None = None
        self._id = 0
        self._lock = threading.Lock()
        self._pending: dict[int, dict[str, Any]] = {}
        self._reader: threading.Thread | None = None
        self._stderr_chunks: list[str] = []

    def __enter__(self) -> McpStdioClient:
        self.start()
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    def start(self) -> None:
        if shutil.which(self.command) is None and not os.path.isfile(self.command):
            raise McpStdioError(f"找不到命令: {self.command}")
        self._proc = subprocess.Popen(
            [self.command, *self.args],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            bufsize=0,
            cwd=self.cwd,
            env=os.environ.copy(),
        )
        self._reader = threading.Thread(target=self._read_loop, daemon=True)
        self._reader.start()
        threading.Thread(target=self._read_stderr, daemon=True).start()
        self.request(
            "initialize",
            {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "workbuddy-ide-review-m0", "version": "0.1.0"},
            },
        )
        self.notify("notifications/initialized", {})

    def close(self) -> None:
        proc = self._proc
        self._proc = None
        if proc is None:
            return
        try:
            if proc.stdin:
                proc.stdin.close()
        except Exception:
            pass
        try:
            proc.terminate()
            proc.wait(timeout=3)
        except Exception:
            try:
                proc.kill()
            except Exception:
                pass

    def notify(self, method: str, params: dict[str, Any] | None = None) -> None:
        msg: dict[str, Any] = {"jsonrpc": "2.0", "method": method}
        if params is not None:
            msg["params"] = params
        self._write(msg)

    def request(self, method: str, params: dict[str, Any] | None = None) -> Any:
        with self._lock:
            self._id += 1
            req_id = self._id
        msg: dict[str, Any] = {"jsonrpc": "2.0", "id": req_id, "method": method}
        if params is not None:
            msg["params"] = params
        event = threading.Event()
        slot: dict[str, Any] = {"event": event, "result": None, "error": None}
        self._pending[req_id] = slot
        self._write(msg)
        if not event.wait(self.timeout_sec):
            self._pending.pop(req_id, None)
            raise McpStdioError(f"MCP 请求超时: {method}")
        if slot["error"] is not None:
            raise McpStdioError(f"MCP 错误 {method}: {slot['error']}")
        return slot["result"]

    def list_tools(self) -> list[dict[str, Any]]:
        result = self.request("tools/list", {})
        if not isinstance(result, dict):
            return []
        tools = result.get("tools") or []
        return tools if isinstance(tools, list) else []

    def call_tool(self, name: str, arguments: dict[str, Any] | None = None) -> Any:
        return self.request(
            "tools/call",
            {"name": name, "arguments": arguments or {}},
        )

    def _write(self, msg: dict[str, Any]) -> None:
        proc = self._proc
        if proc is None or proc.stdin is None:
            raise McpStdioError("MCP 进程未启动")
        body = json.dumps(msg, ensure_ascii=False).encode("utf-8")
        header = f"Content-Length: {len(body)}\r\n\r\n".encode("ascii")
        try:
            proc.stdin.write(header + body)
            proc.stdin.flush()
        except Exception as e:
            raise McpStdioError(f"写入 MCP stdin 失败: {e}") from e

    def _read_loop(self) -> None:
        proc = self._proc
        if proc is None or proc.stdout is None:
            return
        stdout = proc.stdout
        while True:
            msg = self._read_one_message(stdout)
            if msg is None:
                break
            if "id" in msg and ("result" in msg or "error" in msg):
                slot = self._pending.pop(msg["id"], None)
                if slot is None:
                    continue
                if "error" in msg:
                    slot["error"] = msg["error"]
                else:
                    slot["result"] = msg.get("result")
                slot["event"].set()

    def _read_one_message(self, stdout: Any) -> dict[str, Any] | None:
        # Content-Length framing
        headers: dict[str, str] = {}
        while True:
            line = stdout.readline()
            if not line:
                return None
            if line in (b"\r\n", b"\n"):
                break
            try:
                text = line.decode("ascii", errors="replace").strip()
            except Exception:
                continue
            if ":" in text:
                k, v = text.split(":", 1)
                headers[k.strip().lower()] = v.strip()

        if "content-length" in headers:
            try:
                length = int(headers["content-length"])
            except ValueError:
                return None
            raw = stdout.read(length)
            if not raw:
                return None
            try:
                data = json.loads(raw.decode("utf-8"))
            except json.JSONDecodeError:
                return None
            return data if isinstance(data, dict) else None

        return None

    def _read_stderr(self) -> None:
        proc = self._proc
        if proc is None or proc.stderr is None:
            return
        while True:
            line = proc.stderr.readline()
            if not line:
                break
            try:
                self._stderr_chunks.append(line.decode("utf-8", errors="replace"))
            except Exception:
                continue
            if len(self._stderr_chunks) > 50:
                self._stderr_chunks = self._stderr_chunks[-50:]

    def stderr_tail(self) -> str:
        return "".join(self._stderr_chunks[-20:]).strip()


def default_mcp_command_args() -> tuple[str, list[str]]:
    cmd = (os.getenv("IDE_REVIEW_MCP_COMMAND") or "npx").strip()
    raw = (os.getenv("IDE_REVIEW_MCP_ARGS") or "-y,vscode-as-mcp-server").strip()
    args = [a for a in raw.split(",") if a.strip()]
    return cmd, args

File: apps/vscode-workbuddy-bridge/mcpClient.js
/**
 * 最小 MCP stdio 客户端（Content-Length 帧），无第三方依赖。
 * 用于调用 vscode-as-mcp-server 等成熟 MCP 工具（如 code_checker）。
 */
const { spawn } = require("child_process");

class McpStdioClient {
  constructor(command, args, { cwd, timeoutMs } = {}) {
    this.command = command;
    this.args = args || [];
    this.cwd = cwd;
    this.timeoutMs = timeoutMs || 45000;
    this.proc = null;
    this.nextId = 1;
    this.pending = new Map();
    this.buffer = Buffer.alloc(0);
    this.stderrTail = [];
  }

  async start() {
    this.proc = spawn(this.command, this.args, {
      cwd: this.cwd,
      env: process.env,
      stdio: ["pipe", "pipe", "pipe"],
    });
    this.proc.stdout.on("data", (chunk) => this._onStdout(chunk));
    this.proc.stderr.on("data", (chunk) => {
      const s = chunk.toString("utf8");
      this.stderrTail.push(s);
      if (this.stderrTail.length > 40) this.stderrTail.shift();
    });
    this.proc.on("error", (err) => {
      for (const [, slot] of this.pending) {
        slot.reject(err);
      }
      this.pending.clear();
    });

    await this.request("initialize", {
      protocolVersion: "2024-11-05",
      capabilities: {},
      clientInfo: { name: "workbuddy-vscode-bridge", version: "0.2.0" },
    });
    this.notify("notifications/initialized", {});
  }

  close() {
    if (!this.proc) return;
    try {
      this.proc.stdin.end();
    } catch {
      /* ignore */
    }
    try {
      this.proc.kill();
    } catch {
      /* ignore */
    }
    this.proc = null;
  }

  notify(method, params) {
    const msg = { jsonrpc: "2.0", method };
    if (params !== undefined) msg.params = params;
    this._write(msg);
  }

  request(method, params) {
    const id = this.nextId++;
    const msg = { jsonrpc: "2.0", id, method };
    if (params !== undefined) msg.params = params;
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`MCP 超时: ${method}`));
      }, this.timeoutMs);
      this.pending.set(id, {
        resolve: (v) => {
          clearTimeout(timer);
          resolve(v);
        },
        reject: (e) => {
          clearTimeout(timer);
          reject(e);
        },
      });
      try {
        this._write(msg);
      } catch (e) {
        clearTimeout(timer);
        this.pending.delete(id);
        reject(e);
      }
    });
  }

  async listTools() {
    const result = await this.request("tools/list", {});
    return Array.isArray(result?.tools) ? result.tools : [];
  }

  async callTool(name, args) {
    return this.request("tools/call", { name, arguments: args || {} });
  }

  getStderrTail() {
    return this.stderrTail.join("").slice(-2000);
  }

  _write(msg) {
    if (!this.proc || !this.proc.stdin.writable) {
      throw new Error("MCP 进程未启动");
    }
    const body = Buffer.from(JSON.stringify(msg), "utf8");
    const header = Buffer.from(`Content-Length: ${body.length}\r\n\r\n`, "ascii");
    this.proc.stdin.write(Buffer.concat([header, body]));
  }

  _onStdout(chunk) {
    this.buffer = Buffer.concat([this.buffer, chunk]);
    while (true) {
      const headerEnd = this.buffer.indexOf("\r\n\r\n");
      if (headerEnd < 0) return;
      const header = this.buffer.slice(0, headerEnd).toString("ascii");
      const match = /Content-Length:\s*(\d+)/i.exec(header);
      if (!match) {
        this.buffer = this.buffer.slice(headerEnd + 4);
        continue;
      }
      const len = parseInt(match[1], 10);
      const total = headerEnd + 4 + len;
      if (this.buffer.length < total) return;
      const body = this.buffer.slice(headerEnd + 4, total).toString("utf8");
      this.buffer = this.buffer.slice(total);
      let msg;
      try {
        msg = JSON.parse(body);
      } catch {
        continue;
      }
      if (msg && msg.id != null && (msg.result !== undefined || msg.error !== undefined)) {
        const slot = this.pending.get(msg.id);
        if (!slot) continue;
        this.pending.delete(msg.id);
        if (msg.error) slot.reject(new Error(JSON.stringify(msg.error)));
        else slot.resolve(msg.result);
      }
    }
  }
}

/**
 * 把 MCP tools/call 结果尽量规范成 findings[]
 */
function normalizeMcpFindings(result) {
  const findings = [];
  const text = contentText(result);

  const tryList = (list) => {
    if (!Array.isArray(list)) return false;
    let n = 0;
    for (const item of list) {
      if (item && typeof item === "object") {
        findings.push(coerceFinding(item));
        n += 1;
      }
    }
    return n > 0;
  };

  if (result && typeof result === "object") {
    if (tryList(result.findings) || tryList(result.diagnostics) || tryList(result.issues)) {
      return findings;
    }
  }
  if (tryList(result)) return findings;

  const stripped = text.trim();
  if (stripped.startsWith("{") || stripped.startsWith("[")) {
    try {
      return normalizeMcpFindings(JSON.parse(stripped));
    } catch {
      /* ignore */
    }
  }

  // 常见：按行 "path:line: message"
  const lineRe = /^(.+?):(\d+):\d+:\s*(.+)$/;
  for (const line of stripped.split(/\r?\n/)) {
    const m = lineRe.exec(line.trim());
    if (m) {
      findings.push({
        severity: "P1",
        path: m[1],
        line: parseInt(m[2], 10) || 0,
        rule: "mcp",
        message: m[3],
        suggestion: "",
      });
    }
  }
  if (findings.length) return findings;

  if (stripped) {
    findings.push({
      severity: "P2",
      path: "",
      line: 0,
      rule: "mcp-raw",
      message: stripped.slice(0, 1200),
      suggestion: "对照 MCP 原始输出确认",
    });
  }
  return findings;
}

function contentText(result) {
  if (result == null) return "";
  if (typeof result === "string") return result;
  if (typeof result === "object" && Array.isArray(result.content)) {
    return result.content
      .map((b) => (b && b.type === "text" ? b.text : JSON.stringify(b)))
      .join("\n");
  }
  try {
    return JSON.stringify(result);
  } catch {
    return String(result);
  }
}

function coerceFinding(item) {
  let sev = String(item.severity || item.level || item.severityCode || "P2");
  const low = sev.toLowerCase();
  if (["error", "critical", "high", "0"].includes(low)) sev = "P0";
  else if (["warning", "warn", "medium", "1"].includes(low)) sev = "P1";
  else if (["info", "hint", "low", "2", "3"].includes(low)) sev = "P2";
  if (!["P0", "P1", "P2"].includes(sev)) sev = "P2";
  let line = item.line || item.lineno || item.startLine || 0;
  line = parseInt(line, 10) || 0;
  return {
    severity: sev,
    path: String(item.path || item.file || item.uri || item.source || ""),
    line,
    rule: String(item.rule || item.code || item.source || "mcp"),
    message: String(item.message || item.msg || item.text || JSON.stringify(item)),
    suggestion: String(item.suggestion || item.fix || ""),
  };
}

async function runMcpCodeReview({ command, args, cwd, toolName, paths, prompt, timeoutMs }) {
  const client = new McpStdioClient(command, args, { cwd, timeoutMs });
  const meta = { tools: [], toolUsed: "", stderr: "" };
  try {
    await client.start();
    const tools = await client.listTools();
    meta.tools = tools.map((t) => t.name).filter(Boolean);
    const preferred = toolName || "code_checker";
    const pick =
      meta.tools.includes(preferred)
        ? preferred
        : meta.tools.find((n) => /check|diagnostic|lint|review/i.test(n)) || meta.tools[0];
    if (!pick) {
      throw new Error("MCP 未暴露任何 tool");
    }
    meta.toolUsed = pick;

    const argCandidates = [
      { paths: paths || [], prompt: prompt || "" },
      { path: (paths && paths[0]) || "", prompt: prompt || "" },
      {},
    ];
    let lastErr = null;
    let result = null;
    for (const a of argCandidates) {
      try {
        result = await client.callTool(pick, a);
        lastErr = null;
        break;
      } catch (e) {
        lastErr = e;
      }
    }
    if (lastErr && result == null) throw lastErr;

    const findings = normalizeMcpFindings(result);
    return {
      ok: true,
      findings,
      raw_summary: contentText(result).slice(0, 2000),
      meta,
    };
  } catch (e) {
    meta.stderr = client.getStderrTail();
    return {
      ok: false,
      error: String(e.message || e),
      findings: [],
      meta,
    };
  } finally {
    client.close();
  }
}

module.exports = {
  McpStdioClient,
  runMcpCodeReview,
  normalizeMcpFindings,
};
